# Import libraries

In [4]:
import pandas as pd
import numpy as np
import os

# Import Files

In [32]:
#path
path = r'/Users/poojathummar/Documents/career Foundary/Data Immersion/Achi 4 Python Fundamentals for Data Analysts/Instacart Basket Analysis'


In [34]:
#import orders_products_combined_merged_new_variables_derived48.pkl from prepared Data Folder
ords_prods_merge = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_combined_merged_new_variables_derived48.pkl'))


In [35]:
#  to Check dimensions of dataframe for consistency
ords_prods_merge.shape

(32404859, 20)

In [38]:
ords_prods_merge.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2539329        1             1                   2                  8   
2   2539329        1             1                   2                  8   
3   2539329        1             1                   2                  8   
4   2539329        1             1                   2                  8   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                     NaN       14084                  2          0   
2                     NaN       12427                  3          0   
3                     NaN       26088                  4          0   
4                     NaN       26405                  5          0   

                              product_name  aisle_id  department_id  prices  \
0                                     Soda        77              7     9.0   
1  Organic Unsweetened Vanilla Almond Milk        91             16    12.5   
2                      Original Beef Jerky        23             19     4.4   
3               Aged White Cheddar Popcorn        23             19     4.7   
4         XL Pick-A-Size Paper Towel Rolls        54             17     1.0   

  _merge    price_range_loc     busiest_day    busiest_days  \
0   both  Mid-range product  Regularly busy  Regularly busy   
1   both  Mid-range product  Regularly busy  Regularly busy   
2   both  Low-range product  Regularly busy  Regularly busy   
3   both  Low-range product  Regularly busy  Regularly busy   
4   both  Low-range product  Regularly busy  Regularly busy   

  busiest_period_of_day  max_order  loyalty_flag  
0        Average orders         10  New customer  
1        Average orders         10  New customer  
2        Average orders         10  New customer  
3        Average orders         10  New customer  
4        Average orders         10  New customer

# Step 2:

###### In this Exercise, you learned how to find the aggregated mean of the “order_number” column grouped by “department_id” for a subset of your dataframe. Now, repeat this process for the entire dataframe.

In [42]:
ords_prods_merge.groupby('department_id')['order_number'].mean()

department_id
1     15.457838
2     17.277920
3     17.170395
4     17.811403
5     15.215751
6     16.439806
7     17.225802
8     15.340650
9     15.895474
10    20.197148
11    16.170638
12    15.887671
13    16.583536
14    16.773669
15    16.165037
16    17.665606
17    15.694469
18    19.310397
19    17.177343
20    16.473447
21    22.902379
Name: order_number, dtype: float64

 # Step 3:

###### Analyze the result. How do the results for the entire dataframe differ from those of the subset? 

#### <I>Compared to the results from the subset and  the results from the entire dataframe, both includes ALL 'department_id'  but there is only differance in department id by order number , all have increased with entire datafarme.For Instance:in subset department_id 1 has 14.800024,on other hand entire dataframe this values is incresed by 15.457838<I>

# Step 4:

###### Follow the instructions in the Exercise for creating a loyalty flag for existing customers using the transform() and loc() functions.

In [44]:
ords_prods_merge.head(20)

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0    2539329        1             1                   2                  8   
1    2539329        1             1                   2                  8   
2    2539329        1             1                   2                  8   
3    2539329        1             1                   2                  8   
4    2539329        1             1                   2                  8   
5    2398795        1             2                   3                  7   
6    2398795        1             2                   3                  7   
7    2398795        1             2                   3                  7   
8    2398795        1             2                   3                  7   
9    2398795        1             2                   3                  7   
10   2398795        1             2                   3                  7   
11    473747        1             3                   3                 12   
12    473747        1             3                   3                 12   
13    473747        1             3                   3                 12   
14    473747        1             3                   3                 12   
15    473747        1             3                   3                 12   
16   2254736        1             4                   4                  7   
17   2254736        1             4                   4                  7   
18   2254736        1             4                   4                  7   
19   2254736        1             4                   4                  7   

    days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                      NaN         196                  1          0   
1                      NaN       14084                  2          0   
2                      NaN       12427                  3          0   
3                      NaN       26088                  4          0   
4                      NaN       26405                  5          0   
5                     15.0         196                  1          1   
6                     15.0       10258                  2          0   
7                     15.0       12427                  3          1   
8                     15.0       13176                  4          0   
9                     15.0       26088                  5          1   
10                    15.0       13032                  6          0   
11                    21.0         196                  1          1   
12                    21.0       12427                  2          1   
13                    21.0       10258                  3          1   
14                    21.0       25133                  4          0   
15                    21.0       30450                  5          0   
16                    29.0         196                  1          1   
17                    29.0       12427                  2          1   
18                    29.0       10258                  3          1   
19                    29.0       25133                  4          1   

                               product_name  aisle_id  department_id  prices  \
0                                      Soda        77              7     9.0   
1   Organic Unsweetened Vanilla Almond Milk        91             16    12.5   
2                       Original Beef Jerky        23             19     4.4   
3                Aged White Cheddar Popcorn        23             19     4.7   
4          XL Pick-A-Size Paper Towel Rolls        54             17     1.0   
5                                      Soda        77              7     9.0   
6                                Pistachios       117             19     3.0   
7                       Original Beef Jerky        23             19     4.4   
8                    Bag of Organic Bananas        24              4    10.3   
9                Aged White Cheddar Popcorn        2

##### <I> here in the table 'layalty_flag' is craeted.<I>

In [47]:
# Check loyalty flags
ords_prods_merge['loyalty_flag'].value_counts(dropna = False)

loyalty_flag
Regular customer    15876776
Loyal customer      10284093
New customer         6243990
Name: count, dtype: int64

# Step 5:

###### The marketing team at Instacart wants to know whether there’s a difference between the spending habits of the three types of customers you identified. Use the loyalty flag you created and check the basic statistics of the product prices for each loyalty category (Loyal Customer, Regular Customer, and New Customer). What you’re trying to determine is whether the prices of products purchased by loyal customers differ from those purchased by regular or new customers.

In [51]:
ords_prods_merge.groupby('loyalty_flag').agg({'prices': ['mean', 'min', 'max', 'std', 'median']})

prices                                 
                       mean  min      max         std median
loyalty_flag                                                
Loyal customer    10.386336  1.0  99999.0  328.017787    7.4
New customer      13.294670  1.0  99999.0  597.560299    7.4
Regular customer  12.495717  1.0  99999.0  539.720919    7.4

# Step 6:

###### The team now wants to target different types of spenders in their marketing campaigns. This can be achieved by looking at the prices of the items people are buying. Create a spending flag for each user based on the average price across all their orders using the following criteria:

In [55]:
# If the mean of the prices of products purchased by a user is lower than 10, then flag them as a “Low spender.”

In [57]:
# If the mean of the prices of products purchased by a user is higher than or equal to 10, then flag them as
#  a “High spender.”

In [59]:
#  To Create a new column for the average spending of a customer
ords_prods_merge['average_spend'] = ords_prods_merge.groupby(['user_id'])['prices'].transform('mean')

In [61]:
ords_prods_merge.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2539329        1             1                   2                  8   
2   2539329        1             1                   2                  8   
3   2539329        1             1                   2                  8   
4   2539329        1             1                   2                  8   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                     NaN       14084                  2          0   
2                     NaN       12427                  3          0   
3                     NaN       26088                  4          0   
4                     NaN       26405                  5          0   

                              product_name  ...  department_id  prices  \
0                                     Soda  ...              7     9.0   
1  Organic Unsweetened Vanilla Almond Milk  ...             16    12.5   
2                      Original Beef Jerky  ...             19     4.4   
3               Aged White Cheddar Popcorn  ...             19     4.7   
4         XL Pick-A-Size Paper Towel Rolls  ...             17     1.0   

   _merge    price_range_loc     busiest_day    busiest_days  \
0    both  Mid-range product  Regularly busy  Regularly busy   
1    both  Mid-range product  Regularly busy  Regularly busy   
2    both  Low-range product  Regularly busy  Regularly busy   
3    both  Low-range product  Regularly busy  Regularly busy   
4    both  Low-range product  Regularly busy  Regularly busy   

  busiest_period_of_day max_order  loyalty_flag average_spend  
0        Average orders        10  New customer      6.367797  
1        Average orders        10  New customer      6.367797  
2        Average orders        10  New customer      6.367797  
3        Average orders        10  New customer      6.367797  
4        Average orders        10  New customer      6.367797  

[5 rows x 21 columns]

In [63]:
# Assign spender tag to each customer using loc() function
ords_prods_merge.loc[ords_prods_merge['average_spend']<10,'spender_flag']='Low_spender'

In [65]:
ords_prods_merge.loc[ords_prods_merge['average_spend']>= 10, 'spender_flag'] = 'High_spender'

In [67]:
# Check spender flags
ords_prods_merge['spender_flag'].value_counts(dropna = False)

spender_flag
Low_spender     31770614
High_spender      634245
Name: count, dtype: int64

# Step 7:

###### In order to send relevant notifications to users within the app (for instance, asking users if they want to buy the same item again), the Instacart team wants you to determine frequent versus non-frequent customers. Create an order frequency flag that marks the regularity of a user’s ordering behavior according to the median in the “days_since_prior_order” column. The criteria for the flag should be as follows:

###### If the median of “days_since_prior_order” is higher than 20, then the customer should be labeled a “Non-frequent customer.”
###### If the median is higher than 10 and lower than or equal to 20, then the customer should be labeled a “Regular customer.”
###### If the median is lower than or equal to 10, then the customer should be labeled a “Frequent customer.”

In [72]:
# To  Create a new column for customer frequency flags from 'days_since_prior_order' column

ords_prods_merge['customer_frequency'] = ords_prods_merge.groupby(['user_id'])['days_since_prior_order'].transform('median')


In [74]:
ords_prods_merge.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2539329        1             1                   2                  8   
2   2539329        1             1                   2                  8   
3   2539329        1             1                   2                  8   
4   2539329        1             1                   2                  8   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                     NaN       14084                  2          0   
2                     NaN       12427                  3          0   
3                     NaN       26088                  4          0   
4                     NaN       26405                  5          0   

                              product_name  ...  _merge    price_range_loc  \
0                                     Soda  ...    both  Mid-range product   
1  Organic Unsweetened Vanilla Almond Milk  ...    both  Mid-range product   
2                      Original Beef Jerky  ...    both  Low-range product   
3               Aged White Cheddar Popcorn  ...    both  Low-range product   
4         XL Pick-A-Size Paper Towel Rolls  ...    both  Low-range product   

      busiest_day    busiest_days busiest_period_of_day max_order  \
0  Regularly busy  Regularly busy        Average orders        10   
1  Regularly busy  Regularly busy        Average orders        10   
2  Regularly busy  Regularly busy        Average orders        10   
3  Regularly busy  Regularly busy        Average orders        10   
4  Regularly busy  Regularly busy        Average orders        10   

   loyalty_flag average_spend  spender_flag customer_frequency  
0  New customer      6.367797   Low_spender               20.5  
1  New customer      6.367797   Low_spender               20.5  
2  New customer      6.367797   Low_spender               20.5  
3  New customer      6.367797   Low_spender               20.5  
4  New customer      6.367797   Low_spender               20.5  

[5 rows x 23 columns]

In [76]:
ords_prods_merge.loc[ords_prods_merge['customer_frequency'] > 20, 'frequency_flag'] = 'Non-frequent customer'

In [78]:
ords_prods_merge.loc[(ords_prods_merge['customer_frequency'] > 10) & (ords_prods_merge['customer_frequency'] <=20), 'frequency_flag'] = 'Regular customer'

In [80]:
ords_prods_merge.loc[ords_prods_merge['customer_frequency'] <= 10, 'frequency_flag'] = 'Frequent customer'

In [82]:
ords_prods_merge['frequency_flag'].value_counts(dropna = False)

frequency_flag
Frequent customer        21559853
Regular customer          7208564
Non-frequent customer     3636437
NaN                             5
Name: count, dtype: int64

In [91]:
ords_prods_merge[['user_id', 'customer_frequency', 'frequency_flag']].head(50)

user_id  customer_frequency         frequency_flag
0         1                20.5  Non-frequent customer
1         1                20.5  Non-frequent customer
2         1                20.5  Non-frequent customer
3         1                20.5  Non-frequent customer
4         1                20.5  Non-frequent customer
5         1                20.5  Non-frequent customer
6         1                20.5  Non-frequent customer
7         1                20.5  Non-frequent customer
8         1                20.5  Non-frequent customer
9         1                20.5  Non-frequent customer
10        1                20.5  Non-frequent customer
11        1                20.5  Non-frequent customer
12        1                20.5  Non-frequent customer
13        1                20.5  Non-frequent customer
14        1                20.5  Non-frequent customer
15        1                20.5  Non-frequent customer
16        1                20.5  Non-frequent customer
17        1                20.5  Non-frequent customer
18        1                20.5  Non-frequent customer
19        1                20.5  Non-frequent customer
20        1                20.5  Non-frequent customer
21        1                20.5  Non-frequent customer
22        1                20.5  Non-frequent customer
23        1                20.5  Non-frequent customer
24        1                20.5  Non-frequent customer
25        1                20.5  Non-frequent customer
26        1                20.5  Non-frequent customer
27        1                20.5  Non-frequent customer
28        1                20.5  Non-frequent customer
29        1                20.5  Non-frequent customer
30        1                20.5  Non-frequent customer
31        1                20.5  Non-frequent customer
32        1                20.5  Non-frequent customer
33        1                20.5  Non-frequent customer
34        1                20.5  Non-frequent customer
35        1                20.5  Non-frequent customer
36        1                20.5  Non-frequent customer
37        1                20.5  Non-frequent customer
38        1                20.5  Non-frequent customer
39        1                20.5  Non-frequent customer
40        1                20.5  Non-frequent customer
41        1                20.5  Non-frequent customer
42        1                20.5  Non-frequent customer
43        1                20.5  Non-frequent customer
44        1                20.5  Non-frequent customer
45        1                20.5  Non-frequent customer
46        1                20.5  Non-frequent customer
47        1                20.5  Non-frequent customer
48        1                20.5  Non-frequent customer
49        1                20.5  Non-frequent customer

# Step 9:


In [93]:
# To Export the updated dataframe as a pickle file
ords_prods_merge.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'orders_products_combined_merged_new_variables_derived_aggregated2-48.pkl'))
